core/src/main/java/com/linecorp/armeria/server/ServiceErrorHandler.java

```diff
@@ -0,0 +1,154 @@
+ * ServiceErrorHandler errorHandler = (ctx, cause) -> {
+ *     if (cause instanceof IllegalArgumentException) {
+ *         return HttpResponse.of(HttpStatus.BAD_REQUEST);
+ *     }
+ *
+ *     // Return null to let ServerErrorHandler.ofDefault() handle the exception.
+ *     return null;
+ * }
+ *
+ * Server.builder().route().errorHandler(errorHandler)...
+ * }</pre>
+ *
+ * <h2>Recording a service exception (or not)</h2>
+ *
+ * <p>By default, an exception raised by a service or a decorator is captured and recorded into
+ * {@link RequestLog#responseCause()}. You can keep Armeria from recording it while sending the
+ * desired response by returning a failed response whose cause is an {@link HttpStatusException} or
+ * {@link HttpResponseException}:
+ * <pre>{@code
+ * ServerErrorHandler errorHandler = (ctx, cause) -> {
+ *     if (cause instanceof IllegalArgumentException) {
+ *         // IllegalArgumentException is captured into RequestLog#responseCause().
+ *         return HttpResponse.of(HttpStatus.BAD_REQUEST);
+ *     }
+ *
+ *     if (cause instanceof NotFoundException) {
+ *         // NotFoundException is NOT captured into RequestLog#responseCause().
+ *         return HttpResponse.ofFailure(HttpStatusException.of(HttpStatus.NOT_FOUND));
+ *     }
+ *     ...
+ * }
+ * }</pre>
+ *
+ * @see ServerErrorHandler
+ */
+@UnstableApi
+@FunctionalInterface
+public interface ServiceErrorHandler {
+    /**
+     * Returns an {@link HttpResponse} for the given {@link Throwable} raised by a service.
+     * This method is invoked once for each request failed with a service-level exception.
+     *
+     * @param ctx the {@link ServiceRequestContext} of the current request.
+     * @param cause the {@link Throwable} raised by the service that handled the current request.
+     *
+     * @return an {@link HttpResponse} to send to the client, or {@code null} to let the next handler
+     *         specified with {@link #orElse(ServiceErrorHandler)} handle the event.
+     */
+    @Nullable
+    HttpResponse onServiceException(ServiceRequestContext ctx, Throwable cause);
+
+    /**
+     * Returns an {@link AggregatedHttpResponse} generated from the given {@link HttpStatus}, {@code message}
+     * and {@link Throwable}. When {@code null} is returned, the next {@link ServerErrorHandler}
+     * in the invocation chain will be used as a fallback (See {@link #orElse(ServiceErrorHandler)}
+     * for more information).
+     *
+     * @param config the {@link ServiceConfig} that provides the configuration properties.
+     * @param headers the received {@link RequestHeaders}, or {@code null} in case of severe protocol violation.
+     * @param status the desired {@link HttpStatus} of the error response.
+     * @param description an optional human-readable description of the error.
+     * @param cause an optional exception that may contain additional information about the error, such as
+     *              {@link ContentTooLargeException}.
```

Comment:
`ContentTooLargeException` would be not a good example since it is mainly used to limit the content length at HTTP decoders. I am okay to remove `such as ....` part or use `IllegalArgumentException` instead?